### ISSUE TITLE
Add Support for Windows Access Masks to MDtypeAccess

### ISSUE THREAD

**Issue by gwillcox-r7** [rapid7/metasploit-framework#17212]
## Summary
This is something that came up whilst doing code review on https://github.com/rapid7/metasploit-framework/pull/17122 but it would be nice to have something inside `MsDtypeAccess` for the access mask that translates the access mask into a more human readable string.

This would likely involve using some functions along with BinData to form the data parsing functionality.

Spencer mentioned an example of what he might want this to look like:

```
  class MsDtypAccessMask < BinData::Record
    endian :little
    hide   :reserved0, :reserved1
 
    # the protocol field id reserved for protocol-specific access rights
    bit16 :protocol
 
    bit3  :reserved0
    bit1  :sy
    bit1  :wo
    bit1  :wd
    bit1  :rc
    bit1  :de
 
    bit1  :gr
    bit1  :gw
    bit1  :gx
    bit1  :ga
    bit2  :reserved1
    bit1  :ma
    bit1  :as
 
    def bit_names
      names = []
      names << :GENERIC_READ if self.gr
      names << :GENERIC_WRITE if self.gw
      # ... for each of the others
      names
    end
 
    ALL  = MsDtypAccessMask.new({ gr: 1, gw: 1, gx: 1, ga: 1, ma: 1, as: 1, sy: 1, wo: 1, wd: 1, rc: 1, de: 1, protocol: 0xffff })
    NONE = MsDtypAccessMask.new({ gr: 0, gw: 0, gx: 0, ga: 0, ma: 0, as: 0, sy: 0, wo: 0, wd: 0, rc: 0, de: 0, protocol: 0 })
  end
```

This would then allow you to do `"Flags: #{access_mask.bit_names.map(&:to_s).join('|')}"` to retrieve something like `Flags: GENERIC_READ|GENERIC_WRITE`.

This will allow us to more easily parse access mask flags and perform checks against them within our code, paving the way for more advanced checking of permissions on Windows.

EDIT: Forgot to mention but you'll likely want to make use of BinData's `.read` method to read the data in and https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/7a53f60e-e730-4dfe-bbe9-b21b62eb790b for the protocol specification for the ACCESS_MASK field.

# Testing
One can likely do some tests on a sample bit of binary data using IRB and the following code, assuming the code definition above:

```
access_mask = Rex::Proto::MsDtyp::MsDtypAccessMask.read(raw_binary_input_data)
```

Then check that `access_mask` has its fields populated correctly with valid data.

**Comment by JustAnda7** [comment]
Is anyone working on this issue? If not, I would like to contribute to this issue.

**Comment by gwillcox-r7** [comment]
@JustAnda7 Not presently, so feel free to pick it up. This is why we added the `easy` and `newbie-friendly` tags in a hope that people would be curious in trying to work on this 😄 If you have any questions feel free to reach out, our Slack is at https://metasploit.com/slack and my username there is the same as my GitHub handle.

**Comment by JustAnda7** [comment]
What access right does `ma` in `bit1 :ma` and `reserved`correspond to and how to test the changes?

**Comment by gwillcox-r7** [comment]
https://github.com/rapid7/metasploit-framework/blob/ed7d458f07f4cab5b7d2e01d1dd431d5e1667e9a/lib/rex/proto/ms_dtyp.rb is the file you are looking for. The link to map the names to the values should be there as one of the reference links.

Specifically the link is https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/7a53f60e-e730-4dfe-bbe9-b21b62eb790b and the value `ma` you can see on that page.

For testing the changes you will need to find ways of reading access masks that contain these values. The one way to look for this would be to see where the MsDtypAccessMask class is being used. https://github.com/rapid7/metasploit-framework/blob/master/modules/auxiliary/gather/ldap_query.rb will probably be the easiest since you can write a custom LDAP query for an object that contains an access mask containing one of these bits set and then make sure you can read it correctly.

**Comment by JustAnda7** [comment]
What access right does `ma` in `bit1 :ma` and `reserved`correspond to and how to test the changes?

**Comment by JustAnda7** [comment]
Sorry for the delay. I am having some issues with the testing. I will submit a PR once the testing is done correctly.

**Comment by JustAnda7** [comment]
@gwillcox-r7 do you know of any way of testing other than an LDAP query ?

**Comment by gwillcox-r7** [comment]
> @gwillcox-r7 do you know of any way of testing other than an LDAP query ?

You could feed some data into the respective function using `irb` and make sure the resulting data structure is populated correctly. This would take advantage of BinData's `read` function.

Here we would just do `access_mask = Rex::Proto::MsDtyp::MsDtypAccessMask.read(raw_binary_input_data)` and then make sure that the raw binary in `raw_binary_input_data` gets populated into a valid BinData structure and its fields are filled out appropriately according to the https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/7a53f60e-e730-4dfe-bbe9-b21b62eb790b protocol specification.

**Comment by smcintyre-r7** [comment]
Completed in #17288
